File: examples/review_diff_forma.py
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class ReviewDiffInput:
    diff: str
    max_findings: float | None = None

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "ReviewDiffInput":
        return cls(
            diff=data["diff"],
            max_findings=data.get("max_findings"),
        )


@dataclass(frozen=True)
class ReviewDiffFinding:
    path: str
    message: str
    line: float | None = None

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "ReviewDiffFinding":
        return cls(
            path=data["path"],
            message=data["message"],
            line=data.get("line"),
        )


@dataclass(frozen=True)
class ReviewDiffOutput:
    summary: str
    findings: list[ReviewDiffFinding]
    clean: bool

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "ReviewDiffOutput":
        return cls(
            summary=data["summary"],
            findings=[ReviewDiffFinding.from_dict(item) for item in data["findings"]],
            clean=data["clean"],
        )


def assert_review_diff_output(value: Any) -> ReviewDiffOutput:
    data = _assert_review_diff_record(value, "ReviewDiffOutput")
    if "summary" not in data:
        raise ValueError(f'ReviewDiffOutput.summary is required')
    _assert_review_diff_string(data["summary"], "ReviewDiffOutput.summary")
    if "findings" not in data:
        raise ValueError(f'ReviewDiffOutput.findings is required')
    if not isinstance(data["findings"], list):
        raise ValueError(f'ReviewDiffOutput.findings must be a list')
    for index, _ in enumerate(data["findings"]):
        _assert_review_diff_finding(data["findings"][index], f"ReviewDiffOutput.findings[{index}]")
    if "clean" not in data:
        raise ValueError(f'ReviewDiffOutput.clean is required')
    _assert_review_diff_boolean(data["clean"], "ReviewDiffOutput.clean")
    return ReviewDiffOutput.from_dict(data)


def _assert_review_diff_finding(value: Any, path: str) -> ReviewDiffFinding:
    data = _assert_review_diff_record(value, path)
    if "path" not in data:
        raise ValueError(f'{path}.path is required')
    _assert_review_diff_string(data["path"], f"{path}.path")
    if "line" in data and data["line"] is not None:
        _assert_review_diff_number(data["line"], f"{path}.line")
    if "message" not in data:
        raise ValueError(f'{path}.message is required')
    _assert_review_diff_string(data["message"], f"{path}.message")
    return ReviewDiffFinding.from_dict(data)


def _assert_review_diff_record(value: Any, path: str) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise ValueError(f"{path} must be an object")
    return value


def _assert_review_diff_string(value: Any, path: str) -> None:
    if not isinstance(value, str):
        raise ValueError(f"{path} must be a string")


def _assert_review_diff_number(value: Any, path: str) -> None:
    if not isinstance(value, (int, float)) or isinstance(value, bool):
        raise ValueError(f"{path} must be a number")


def _assert_review_diff_boolean(value: Any, path: str) -> None:
    if not isinstance(value, bool):
        raise ValueError(f"{path} must be a boolean")

File: examples/review-diff-agent/src/review_diff.forma.ts
export interface ReviewDiffInput {
  diff: string;
  max_findings?: number;
}

export interface ReviewDiffOutput {
  summary: string;
  findings: ReviewDiffFinding[];
  clean: boolean;
}

export interface ReviewDiffFinding {
  path: string;
  line?: number;
  message: string;
}
export function assertReviewDiffOutput(value: unknown): ReviewDiffOutput {
  const data = assertReviewDiffRecord(value, "ReviewDiffOutput");
  assertReviewDiffString(data.summary, "ReviewDiffOutput.summary");
  if (!Array.isArray(data.findings)) throw new Error(`ReviewDiffOutput.findings must be an array`);
  (data.findings as unknown[]).forEach((_, index) => {
    assertReviewDiffFinding((data.findings as unknown[])[index], `ReviewDiffOutput.findings[${index}]`);
  });
  assertReviewDiffBoolean(data.clean, "ReviewDiffOutput.clean");
  return data as unknown as ReviewDiffOutput;
}

function assertReviewDiffFinding(value: unknown, path: string): ReviewDiffFinding {
  const data = assertReviewDiffRecord(value, path);
  assertReviewDiffString(data.path, `${path}.path`);
  if (data.line !== undefined) {
    assertReviewDiffNumber(data.line, `${path}.line`);
  }
  assertReviewDiffString(data.message, `${path}.message`);
  return data as unknown as ReviewDiffFinding;
}

function assertReviewDiffRecord(value: unknown, path: string): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) throw new Error(`${path} must be an object`);
  return value as Record<string, unknown>;
}

function assertReviewDiffString(value: unknown, path: string): void {
  if (typeof value !== "string") throw new Error(`${path} must be a string`);
}

function assertReviewDiffNumber(value: unknown, path: string): void {
  if (typeof value !== "number") throw new Error(`${path} must be a number`);
}

function assertReviewDiffBoolean(value: unknown, path: string): void {
  if (typeof value !== "boolean") throw new Error(`${path} must be a boolean`);
}
